src/types/src/idl.y

```diff
@@ -0,0 +1,477 @@
+  ;
+members:
+    member members
+  | member
+  ;
+
+member:
+    type_spec { dds_ts_add_struct_member(context, $1); }
+    declarators ';'
+  ;
+
+struct_forward_dcl:
+    "struct" identifier
+      { dds_ts_add_struct_forward(context, $2); };
+
+array_declarator:
+    identifier
+    fixed_array_sizes { dds_ts_add_declarator(context, $1); }
+  ;
+
+fixed_array_sizes:
+    fixed_array_size fixed_array_sizes
+  | fixed_array_size
+  ;
+
+fixed_array_size:
+    '[' positive_int_const ']' { dds_ts_add_array_size(context, &($2)); }
+  ;
+
+simple_declarator: identifier ;
+
+declarators:
+    declarator ',' declarators
+  | declarator
+  ;
+
+declarator: simple_declarator
+      { dds_ts_add_declarator(context, $1); };
+
+
+/* From Building Block Extended Data-Types: */
+struct_def:
+    "struct" identifier ':' scoped_name '{'
+      { dds_ts_add_struct_extension_open(context, $2, $4); }
+    members '}'
+      { dds_ts_struct_close(context); }
+  | "struct" identifier '{'
+      { dds_ts_add_struct_open(context, $2); }
+    '}'
+      { dds_ts_struct_empty_close(context); }
+  ;
+
+template_type_spec:
+     map_type
+  ;
+
+map_type:
+    "map" '<' type_spec ',' type_spec ',' positive_int_const '>'
+      { $$ = dds_ts_new_map_type(context, $3, $5, &($7)); }
+  | "map" '<' type_spec ',' type_spec '>'
+      { $$ = dds_ts_new_map_type_unbound(context, $3, $5); }
+  ;
+
+signed_int:
+    signed_tiny_int
+  | signed_short_int
+  | signed_long_int
+  | signed_longlong_int
+  ;
+
+unsigned_int:
+    unsigned_tiny_int
+  | unsigned_short_int
+  | unsigned_long_int
+  | unsigned_longlong_int
+  ;
+
+signed_tiny_int: "int8" { $$ = DDS_TS_INT8_TYPE; };
+unsigned_tiny_int: "uint8" { $$ = DDS_TS_UINT8_TYPE; };
+signed_short_int: "int16" { $$ = DDS_TS_SHORT_TYPE; };
+signed_long_int: "int32" { $$ = DDS_TS_LONG_TYPE; };
+signed_longlong_int: "int64" { $$ = DDS_TS_LONG_LONG_TYPE; };
+unsigned_short_int: "uint16" { $$ = DDS_TS_UNSIGNED_SHORT_TYPE; };
+unsigned_long_int: "uint32" { $$ = DDS_TS_UNSIGNED_LONG_TYPE; };
+unsigned_longlong_int: "uint64" { $$ = DDS_TS_UNSIGNED_LONG_LONG_TYPE; };
+
+/* From Building Block Anonymous Types: */
+type_spec: template_type_spec ;
+declarator: array_declarator ;
+
+
+identifier:
+    IDENTIFIER
+      {
+        size_t offset = 0;
+        if ($1[0] == '_') {
+          offset = 1;
+        } else if (illegale_identifier($1) != 0) {
+          yyerror(&yylloc, scanner, context, "Identifier matches a keyword");
+          YYABORT;
+        }
+        if (($$ = strdup($1 + offset)) == NULL) {
+          YYABORT;
+        }
+      };
+
+/* Backward compatability #pragma keylist */
+
+
+%%
+
+int
+yyerror(
+  YYLTYPE *yylloc, yyscan_t yyscanner, dds_ts_context_t *context, char *text)
+{
+  (void)yyscanner;
+  if (dds_ts_context_get_ignore_yyerror(context)) {
+    return 0;
+  }
+
+  fprintf(stderr, "ERROR: %d.%d %s\n", yylloc->first_line, yylloc->first_column, text);
+  return 0;
+}
+
+int
+illegale_identifier(const char *token)
```

Comment:
This function implements the checks as described in Section 7.2.4 (of the IDL 4.2 specification docment) and the escaping mechanism as mentioned in Section 7.2.3.2. Section 7.2.4 mentions that identifiers like Boolean and BOOLEAN are illegal, because the collide with the keyword boolean.

Reply:
I think that the lexer shouldn't return that it's an identifier if it's not(?). Unless it complicates the parser logic, or you need it to generate a specialized error message. Both are not the case(?). Apart from that, I think you may have intended to write illegal as opposed to illegale(?).

Reply:
An illegal identifier could be viewed as a lexical error or a semantic error. With respect to parsing, there is no problem to regard it as a semantic error, and just proceed parsing the remainder of the input. The question is also whether you want parsing to stop at the first error or report as much errors as possible. Just like an identifier can 'collide' with a keyword, it can also 'collide' with another identifier. ('collide' means, according to the standard, that they are not the same when taking into account case, but are when compared case insensitive.) When we stop parsing on an illegal identifier, maybe we should also stop parsing when we find an colliding identifier.